### PR TITLE
[r] Pass config through `shape()` and `nnz()`

### PR DESCRIPTION
**Pull request by mlin**
**Issue and/or context:** #1137

**Changes:** Factor out a common helper for converting the config character vector to a std::map suitable for `SOMAReader`.

**Notes for Reviewer:** I've tested this in our R CellCensus package, which needs this for `shape()` to read from the correct S3 endpoint. It's tricky to unit-test though, suggestions welcome

### CODE REVIEW

**Review by eddelbuettel — APPROVED** [review]
Looks good and is very helpful.  I have one suggested simplification we can probably introduce now that the config vector is no longer optional.

**Comment by eddelbuettel** [comment]
One thing that would be nice though is _a_ test with a config set this way.  Looking at `tiledb::config( tiledb::tiledb_get_context() )` and its long list maybe turning the logging format on and off.  The real litmus test is of course that it lets you e.g. pass the AWS parameter through as it should.

File: apis/r/src/rutilities.h [hunk 1]
```diff
@@ -64,3 +64,20 @@ void apply_dim_ranges(
...
+        Rcpp::CharacterVector namesvec = confvec.attr("names"); // extract names from named R vector
+        size_t n = confvec.length();
+        for (size_t i = 0; i<n; i++) {
+            platform_config.emplace(std::make_pair(std::string(namesvec[i]), std::string(confvec[i])));
+            spdl::debug("[config_vector_to_map] adding '{}' = '{}'", std::string(namesvec[i]), std::string(confvec[i]));
+        }
+    }
+
+    return platform_config;
+}
```
[the remark this answers is not in the document]
Comment (eddelbuettel) [reply apis/r/src/rutilities.h:83]: Nope. See below.

**Comment by eddelbuettel** [comment]
What is a bit of a wart though is that despite us doing all the work for an _optional_ config, it is now forced at the top:

```
> library(tiledbsoma)
> uri <- "/tmp/tiledb/penguins"
> str(nnz)
function (uri, config)  
> nnz(uri, NULL)
Error: Not compatible with STRSXP: [type=NULL].
> nnz(uri, character())
Error: Not compatible with STRSXP: [type=NULL].
> nnz(uri, c(NONE=""))
[1] 344
> 
```

I think we can do better by keeping this optional just as `some_reader()` does.

```r
> library(tiledbsoma)
> uri <- "/tmp/tiledb/penguins"
> str(nnz)
function (uri, config = NULL)  
> nnz(uri)
[1] 344
> 
```

Much better. It is a simple one-line change to the signature (plus a run of `Rcpp::compileAttributes()`) as you factored out the treatment of the NULL into the setter of the map.

**Comment by eddelbuettel** [comment]
Ditto for shape. I have that working here and can add the simple commit.

```sh
edd@ahmad:~/git/tiledb-soma/apis/r(mlin/r-config-for-shape-and-nnz)$ Rscript -e 'tiledbsoma::shape("/tmp/tiledb/penguins")'
integer64
[1] 344
edd@ahmad:~/git/tiledb-soma/apis/r(mlin/r-config-for-shape-and-nnz)$ 
```

**Comment by eddelbuettel** [comment]
As I wrote, I have the (small) extension to make config optional working.  Do you want to commit this?  It tests cleanly here.  Of course, if you would rather (re-)do it yourself that is perfectly fine too.

I didn't yet get to a unit test.  My idea was to maybe turn logging on/off and assert stdout.

**Comment by mlin** [comment]
@eddelbuettel Sure, I welcome your input on both items of course

**Comment by eddelbuettel** [comment]
Hold on, I just realized I should be a good boy and also mark the config as optional and update the Rd files.  One sec.

**Comment by eddelbuettel** [comment]
From a quick Slack discussion with the core team, one way to 'break' access is to set a borked key and being not exactly truthful about encryption (by pretending there was some on a quick local array).  

```r
> tiledbsoma::nnz("/tmp/tiledb/penguins", c(sm.encryption_key="Nope", sm.encryption_type="AES_256_GCM"))
Error: Error opening array: '/tmp/tiledb/penguins'
  [TileDB::Array] Error: [TileDB::Encryption] Error: Cannot create key; invalid key length for encryption type.
> 
```

**Comment by eddelbuettel** [comment]
I added tests for `nnz()` and `shape()` based on the idea above.

**Review by johnkerl — APPROVED** [review]
🔥

**Pull request merged**